### PR TITLE
Enable dynamically loading BT nodes

### PR DESCRIPTION
**Pull request by orduno**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #942 |
| Primary OS tested on | Ubuntu 18.04 |
| Robotic platform tested on | Gazebo simulation of TB3 |

---

## Description of contribution in a few bullet points

Addressing #942:

* Making the existing nav2 behavior tree nodes plugins.
* Modifying BT engine to load plugin library specified on params file.

---

## Future work that may be required in bullet points

Provide documentation explaining users how to create and use a navigation behavior tree with custom nodes.

### CODE REVIEW

**Comment by SteveMacenski** [comment]
Shouldn’t all the nodes be separate plugins?

**Comment by crdelsey** [comment]
> Shouldn’t all the nodes be separate plugins?

They could be. The behavior tree library has an example done this way though.

I figured it would be more convenient for users to have just a single item to load to get the full set of nav2 bt nodes, and I don't see much downside. Also, some nodes can be as small as a single function. Breaking them into a separate library seems  like overkill.

File: nav2_behavior_tree/src/behavior_tree_engine.cpp [hunk 1]
```diff
@@ -16,51 +16,20 @@
...
-  factory_.registerSimpleCondition("initialPoseReceived",
-    std::bind(&BehaviorTreeEngine::initialPoseReceived, this, std::placeholders::_1));
-
-  // Register our custom decorator nodes
-  factory_.registerNodeType<nav2_behavior_tree::RateController>("RateController");
-
-  // Register our custom control nodes
-  factory_.registerNodeType<nav2_behavior_tree::RecoveryNode>("RecoveryNode");
+  for (const auto & p : plugin_libraries) {
+    factory_.registerFromPlugin(std::string{"lib" + p + ".so"});
```
Comment (crdelsey) [inline nav2_behavior_tree/src/behavior_tree_engine.cpp]: This isn't going to work for non-linux systems. We need to find a way to make this cross-platform.

Maybe you can take a look at the pluginlib/class loader code to see how they make it cross-platform

Comment (orduno) [reply nav2_behavior_tree/src/behavior_tree_engine.cpp]: Right, that seemed something to be addressed upstream.

In the meantime, we can try what [pluginlib does](https://github.com/ros/pluginlib/blob/37e1763edb988e3c4e07de76a4e22163a81b5219/pluginlib/include/pluginlib/class_loader_imp.hpp#L377) which is to get a list of all possible paths and filename combinations to try...

**Comment by SteveMacenski** [comment]
I think making them separate is important so they’re actually plugins and each can be chosen. Right now if you didn’t want a spin server or recovery, you don’t have that option.

**Comment by orduno** [comment]
On a single library model, all nav2 nodes are loaded and the bt xml description defines which are actually used. No problem if there are unused nodes. Ok to load all since these are relatively small. However, there might be some undefined behavior if the user loads nav2 nodes, and then a custom library, both of which implement a `FollowPath` node. This could be solved by adding some namespacing like `nav2::FollowPath`, not a problem on multi-library model.

The main advantage I see to having them on separate libraries is that the user would clearly what node types are available on the params file.

I'm ok going either way.

**Comment by SteveMacenski** [comment]
> However, there might be some undefined behavior if the user loads nav2 nodes, and then a custom library, both of which register a node with a FollowPath name

Why would a user do that? I don't think that’s a concern. This is true of literally all plugins, if you register twice on one system, that’s bad. I don’t find that a convincing argument.

**Comment by orduno** [comment]
I'm pointing out that that is a problem with having a single nav2 nodes lib, therefore, in that case, it's better to go with separate libs (if we don't want to add namespacing complexity).

I'll push a commit separating the plugins.

**Comment by crdelsey** [comment]
>  think making them separate is important so they’re actually plugins and each can be chosen. Right now if you didn’t want a spin server or recovery, you don’t have that option.

If you don't add the spin action to your behavior tree XML file, you don't have a spin recovery; you just have some unused code loaded into RAM. It's kind of like linking with the standard library; printf gets loaded but you don't have to use it, and it just sits there.

I'm OK with splitting them up if that is really the way everyone wants to go, but we've got on the order of 13 different custom nodes to load with probably more coming down the line. That seems like a pain to have to list out all the time. It seems easier to me to instead say, I want all the nav2 nodes available and leave it up to the XML to specify which are used.

**Comment by SteveMacenski** [comment]
Also keep in mind this is going to be the single, defacto, and distributed code for the BT navigator. 

If people want to make their own plugin nodes, separating them gives them plenty of examples to learn from. Otherwise 1 massive plugin obscures what needs to be done to enable their application. Its obvious to us because we understand it, but its certainly not obvious to most, especially since a BT by itself is a novel concept to most in the robotics community. 

Even if not 13 plugins but N, where theyre sorted into themes would help so there are multiple examples. All the recoveries in 1, etc

**Comment by crdelsey** [comment]
Ok. @orduno Go ahead and split it into multiple libs.

**Comment by orduno** [comment]
@SteveMacenski I've split the nodes into separate libraries, can you take a look?

**Comment by SteveMacenski** [comment]
I wish I had a chance to review this before going in but it looks fine.

The only complaint I have is that it kind of defeats the purpose of plugins if they're not included in their implementations. Why is the plugin factory definition **not** in the header file for the node? That would make alot of sense to me to be analogous for how pluginlib works and encapsulates all the items in one file. 

Also then when/if we make those headers pluginlib plugins then its a one shot file to register plugins across the board.

**Comment by SteveMacenski** [comment]
@orduno thoughts on that?

**Comment by orduno** [comment]
Sure, we could just have a `.cpp` file for the plugins.  I was following the pattern used on the bt library examples.

**Pull request merged**